### ISSUE TITLE
shadow master fails to (re-)start after abnormal termination

### ISSUE THREAD

**Issue by onlyjob** [lizardfs/lizardfs#284]
Power surge rebooted shadow master. After reboot shadow master (2.6.0) failed to start:

~~~~
May 29 00:29:39 deblab mfsmaster[1856]: calculating checksum of the metadata
May 29 00:29:45 deblab mfsmaster[1856]: metadata file /var/lib/lizardfs/metadata/metadata.mfs read (28571217 inodes including 1511321 directory inodes and 26873268 file inodes, 6039852 chunks)
May 29 00:29:45 deblab mfsmaster[1856]: running in shadow mode - applying changelogs from /var/lib/lizardfs/metadata
May 29 00:29:46 deblab mfsmaster[1856]: /var/lib/lizardfs/metadata/changelog.mfs.1: 231774 changes applied (2056346476 to 2056578249), 0 skipped
May 29 00:29:46 deblab mfsmaster[1856]: terminate called after throwing an instance of 'std::invalid_argument'
May 29 00:29:46 deblab mfsmaster[1856]: what():  stoull
~~~~

I had to delete `changelog.mfs.*` to fix the problem.

It would be nice to automatically dismiss changelog with failed CRC and re-download it from master.
Ideally shadow master should be able to initialise by re-downloading data from master as necessary.

**Comment by onlyjob** [comment]
Just reproduced the issue: unclean shutdown leaves `changelog.mfs` corrupted. It is sufficient to remove this file to allow shadow master to start. Perhaps this could be done automatically?

**Comment by psarna** [comment]
`changelog.mfs` may hold some valuable data, even if it seems corrupted - usually most of it is still usable, except for few last lines. If anything should be done automatically, it should be restoring `changelog.mfs` by removing incomplete lines - we will consider implementing such a tool once there is time.

**Comment by onlyjob** [comment]
Thank you. Anything you could do in order to help shadow master to start without manual admin action will be useful.

**Comment by DarkHaze** [comment]
On the one hand we can do change in the way master reads data from the change log. 
And probably we will do it. But I'm starting to wonder if we should. 

The problem now is that at the end of changelog file there are garbage data caused by unclean system shutdown/failure. This can be fixed by enabling full journalling in the ext4 file system ( it's only required for partition holding metadata ). This can be done by using data=journal option in mount. Of course it's much slower than default ordered mode.

So here is what bothers me. I think that maybe enabling full journalling for master config/metadata should be required. And the fix for this problem should be checking at start if this is enabled and warning if it's not.

**Comment by onlyjob** [comment]
No, let's not make assumptions regarding file system type and the way how it handles power cut-off.
By definition do not put your trust to file system or hardware. Ever. Journaled mode or not. File system may have bugs; hardware may have bugs. Corruption happens -- this is just a fact of life.

The problem is _not_ that corruption happened or how it happened. The problem is that shadow master fails when it should have been able to continue by re-fetching data from master. Data is lost, gone. Does not matter how. And it works when corrupted file deleted manually. So just add exception handler to dismiss unusable data in shadow mode. Even if partially usable, would you trust such data anyway?

**Comment by DarkHaze** [comment]
The solution we will do will add extra crc to each committed line ( = operation ) in the changle log file.  So I'm guessing you can trust it. And using this data will give more up to date metadata version in newly started master.

Of course I can't guarantee now that it is exactly how we will fix it. 

But it's my personal opinion that if possible then full journalling should be enabled in master anyway because it will save master from many other strange and hairy scenarios I can think of ( and some of them probably can't be solved by other means ).

**Comment by onlyjob** [comment]
Thank you. Adding more CRC (i.e. integrity) checks would be a correct approach. It might prevent very unpleasant issues with metadata corruption.

It might be useful to enable journaling (if `ext4` is in use) but this particular case is with _shadow_ master. Our primary master is protected by UPS but even so there might be kernel OOPS or hardware malfunction where full ext4 journaling won't be helpful...

**Comment by onlyjob** [comment]
7 months later I'm still hoping for this improvement to happen...
Perhaps LizardFS team should consider organising a bug squashing party? ;-)

**Comment by jwest75674** [comment]
Years later, and as @onlyjob predicted, a kernel oops has me unable to start my master due to corrupt a changelog having garbage at the tail end. Shadow starts fine, but haven't yet figured out how to start the master... Deleting the junk leads to a "change log has holes" error...

**Comment by jwest75674** [comment]
Ended up getting the shadow upgraded, but it was quite a hassle. From there, with roles reversed, recovery fell into place.
